src/utils/transaction.ts

```diff
@@ -572,6 +574,124 @@ const transaction = {
     throw new ParseError('Invalid transaction version.');
   },
 
+  /**
+   * Convert a Transaction instance to the history object.
+   * May call the fullnode transaction api to get information on the tx spent
+   * by the inputs.
+   */
+  async convertTransactionToHistoryTx(
+    tx: Transaction | CreateTokenTransaction | NanoContract,
+    storage: IStorage
+  ): Promise<IHistoryTx> {
+    if (!tx.hash) {
+      throw new Error('To be a history tx a calculated hash is required');
+    }
+
+    const inputs: IHistoryInput[] = [];
+    const outputs: IHistoryOutput[] = [];
+    const txCache: Record<string, IHistoryTx> = {};
+
+    for (const input of tx.inputs) {
+      let spentTx = await storage.getTx(input.hash);
+      if (!spentTx) {
+        // Try cache first
+        if (txCache[input.hash]) {
+          spentTx = txCache[input.hash];
+        } else {
+          // Get from API
+          spentTx = await new Promise((resolve, reject) => {
+            txApi
+              .getTransaction(input.hash, (response: FullNodeTxApiResponse) => {
+                if (!response.success) {
+                  return reject(new Error(response.message ?? ''));
+                }
+
+                if (input.index >= response.tx.outputs.length) {
+                  return reject(new Error('Index outside of tx output array bounds'));
+                }
+                return resolve(this.convertFullNodeTxToHistoryTx(response));
+              })
+              .catch(err => reject(err));
+          });
+
+          if (!spentTx) {
+            // This should not happen since any errors should be treated already.
+            // This if statement is to ensure typing since spentTx starts as IHistoryTx | null.
+            throw new Error('Could not find the spent transaction');
+          }
+          // Update cache
+          txCache[input.hash] = cloneDeep(spentTx);
+        }
+      }
+
+      if (input.index >= spentTx.outputs.length) {
+        throw new Error(
+          `Index (${input.index}) outside of transaction output array bounds (${spentTx.outputs.length})`
+        );
+      }
+
+      const spentOut = spentTx.outputs[input.index];
+      inputs.push({
+        tx_id: input.hash,
+        index: input.index,
+        script: spentOut.script,
+        decoded: spentOut.decoded,
+        token_data: spentOut.token_data,
```

Comment:
I think this might be wrong. The tokens array of the spent tx might be different than the tokens array of the tx, so the token data might be different.

Reply:
Yes, but the `token_data` of the input is the `token_data` on the transaction being spent (checked on the hathor-code repo).

The token array is only for the tokens on the outputs, for instance, if we have a tx that is melting tokens and not creating a melt authority the token array will be empty, meaning that the token_data for the token being melted cannot be constructed with the array from the current tx.